sql/src/main/java/io/crate/planner/consumer/ManyTableConsumer.java

```diff
@@ -93,37 +95,46 @@ public Plan consume(AnalyzedRelation relation, ConsumerContext context) {
      * allows the most join condition push downs (assuming that a left-based tree is built later on)
      *
      * @param relations               all relations, e.g. [t1, t2, t3, t3]
-     * @param implicitJoinedRelations contains all relations that have a join condition e.g. {{t1, t2}, {t2, t3}}
+     * @param explicitJoinedRelations contains all relations that have an explicit join condition
+     * @param implicitJoinedRelations contains all relations that have an implicit join condition
+     *                                e.g. {{t1, t2}, {t2, t3}}
      * @param joinPairs               contains a list of {@link JoinPair}.
      * @param preSorted               a ordered subset of the relations. The result will start with those relations.
      *                                E.g. [t3] - This would cause the result to start with [t3]
      */
     static Collection<QualifiedName> orderByJoinConditions(Collection<QualifiedName> relations,
+                                                           Set<? extends Set<QualifiedName>> explicitJoinedRelations,
                                                            Set<? extends Set<QualifiedName>> implicitJoinedRelations,
                                                            List<JoinPair> joinPairs,
                                                            Collection<QualifiedName> preSorted) {
         if (relations.size() == preSorted.size()) {
             return preSorted;
         }
-        if (relations.size() == 2 || (joinPairs.isEmpty() && implicitJoinedRelations.isEmpty())) {
+        if (relations.size() == 2 ||
+            (joinPairs.isEmpty() && explicitJoinedRelations.isEmpty() && implicitJoinedRelations.isEmpty())) {
             LinkedHashSet<QualifiedName> qualifiedNames = new LinkedHashSet<>(preSorted);
             qualifiedNames.addAll(relations);
             return qualifiedNames;
         }
+        if (preSorted.isEmpty()) {
+            Set<QualifiedName> qualifiedNames = new LinkedHashSet<>(preSorted);
+            qualifiedNames.addAll(explicitJoinedRelations.stream().flatMap(Collection::stream).collect(Collectors.toList()));
```

Comment:
which intermediate list are you referring to?

We have a linkedHashSet<linkedHashet<QualifiedName>> and we want to flatten it and also keep the order.